crypto/bls12381/key_bls12381.go

```diff
@@ -0,0 +1,159 @@
+	PubKeySize = 48
+	// SignatureLength defines the byte length of a BLS signature.
+	SignatureLength = 96
+	// KeyType is the string constant for the bls12_381 algorithm.
+	KeyType = "bls12_381"
+	// Enabled indicates if this curve is enabled.
+	Enabled = true
+)
+
+// -------------------------------------.
+const (
+	PrivKeyName = "cometbft/PrivKeyBLS12_381"
+	PubKeyName  = "cometbft/PubKeyBLS12_381"
+)
+
+func init() {
+	cmtjson.RegisterType(PubKey{}, PubKeyName)
+	cmtjson.RegisterType(PrivKey{}, PrivKeyName)
+}
+
+// ===============================================================================================
+// Private Key
+// ===============================================================================================
+
+// PrivKey is a wrapper around the Ethereum bls12_381 private key type. This
+// wrapper conforms to crypto.Pubkey to allow for the use of the Ethereum
+// bls12_381 private key type.
+
+var _ crypto.PrivKey = &PrivKey{}
+
+type PrivKey []byte
+
+func NewPrivateKeyFromBytes(bz []byte) (PrivKey, error) {
+	secretKey, err := bls12381.SecretKeyFromBytes(bz)
+	if err != nil {
+		return nil, err
+	}
+	return secretKey.Marshal(), nil
+}
+
+func GenPrivKey() (PrivKey, error) {
+	secretKey, err := bls12381.RandKey()
+	return PrivKey(secretKey.Marshal()), err
+}
+
+// Bytes returns the byte representation of the Private Key.
+func (privKey PrivKey) Bytes() []byte {
+	return privKey
+}
+
+// PubKey returns the private key's public key. If the privkey is not valid
+// it returns a nil value.
+func (privKey PrivKey) PubKey() crypto.PubKey {
+	secretKey, _ := bls12381.SecretKeyFromBytes(privKey)
+
+	return PubKey(secretKey.PublicKey().Marshal())
+}
+
+// Equals returns true if two private keys are equal and false otherwise.
+func (privKey PrivKey) Equals(other crypto.PrivKey) bool {
+	return privKey.Type() == other.Type() && bytes.Equal(privKey.Bytes(), other.Bytes())
+}
+
+// Type returns eth_bls12_381.
+func (PrivKey) Type() string {
+	return KeyType
+}
+
+func (privKey PrivKey) Sign(digestBz []byte) ([]byte, error) {
+	secretKey, err := bls12381.SecretKeyFromBytes(privKey)
+	if err != nil {
+		return nil, err
+	}
+
+	bz := digestBz
+	if len(bz) > 32 {
+		hash := sha256.Sum256(bz)
+		bz = hash[:]
+	}
+
+	sig := secretKey.Sign(bz)
+	return sig.Marshal(), nil
+}
+
+// ===============================================================================================
+// Public Key
+// ===============================================================================================
+
+// Pubkey is a wrapper around the Ethereum bls12_381 public key type. This
+// wrapper conforms to crypto.Pubkey to allow for the use of the Ethereum
+// bls12_381 public key type.
+
+var _ crypto.PubKey = &PubKey{}
+
+type PubKey []byte
+
+// Address returns the address of the public key.
+// The function will panic if the public key is invalid.
+func (pubKey PubKey) Address() crypto.Address {
+	pk, _ := bls12381.PublicKeyFromBytes(pubKey)
```

Comment:
Since we're panicking in line 122, can we please check the error here and have a dedicated panic? Otherwise, we may have a `panic` in line 121 when trying to marshall a `nil` public key.